송원/후위표기식1/index.py

```diff
@@ -0,0 +1,40 @@
+import sys
+sys.stdin = open("input.txt", "rt")
+
+notation = input()
+priority = {
+  '(': 0,
+  '+': 1,
+  '-': 1,
+  '*': 2,
+  '/': 2
+}
+
+answer = ''
+stack = []
+
+for i in notation:
+  if i.isalpha():
+    answer += i
+  else:
+    if len(stack) == 0:
+      stack.append(i)
+    else:
+      if i == ')':
+        while stack[-1] != '(':
+          answer += stack.pop()
+        stack.pop()
+
+      elif i == '(':
+        stack.append(i)
+      else:
+        while priority[stack[-1]] >= priority[i]:
+          answer += stack.pop()
+          if len(stack) == 0:
+            break
+        stack.append(i)
```

Comment:
```python
if len(stack) == 0:
  stack.append(i)
else:
  # ...
```
제 생각에는 `stack`의 length을 체크하여 depth가 더 들어가는 것 보다는
`pop()`을 수행할 때 마다 length을 체크하는 것은 어떨까요?

문제에서 <b>-A+B와 같이 -가 가장 앞에 오거나 AB와 같이 *가 생략되는 등의 수식은 주어지지 않는다.</b>라고
적혀있는 것을 봤을 때 **연산자**가 **입력의 맨 앞에 오는 경우가 없다**는 것을 알려주고 있으며
이 말은 **문자** or **여는 괄호**가 문자의 맨 앞에 올 수 있다는 것을 뜻한다고 생각합니다.

결과적으로 송원님의 코드를 위의 suggestion 코드와 합쳐서 다음처럼 줄일 수 있을 것 같습니다.
(돌려보지는 않아서.. 틀릴 수 있습니다..)

```suggestion
  if i.isalpha():
    answer += i
  elif i == ')':
    while len(stack) > 0 and stack[-1] != '(':
      answer += stack.pop()
    stack.pop()
  elif i == '(':
    stack.append(i)
  else:
    while len(stack) > 0 and priority[stack[-1]] >= priority[i]:
      answer += stack.pop()
    stack.append(i)
```

Reply:
아 문제를 잘 봐야 코드도 줄일 수 있는거군요...! 그냥 일단 답 구하는게 정답은 아니었네요